### PR TITLE
Implement the procedure for calculating an auto-aligned start time

### PR DESCRIPTION
**Pull request by johannesodland**
Currently animation range start and range end is mapped to delay and endDelay of the animation effect. This causes the finished promise to resolve at the wrong time, and similarly the finished event is emitted at the wrong time.

This PR implements the procedure for calculating an auto-aligned start time.  (https://github.com/w3c/csswg-drafts/pull/9181/files). It aligns the start time based on animation range start, no longer maps ranges to effect delay and end delay.

<details>
<summary>Summary of changes to proxy-animation:</summary>

- Remove the reset current time on resume flag  (https://github.com/w3c/csswg-drafts/pull/9181/files)
- Update the conditions for when an animation is ready (https://drafts.csswg.org/web-animations-2/#waiting-for-the-associated-effect)
- Update timeline, startTime and currentTime setters so that they sets the `auto align start time` flag
- Update playInternal so that 
  - it sets the `auto align start time` flag
  - it delays add animation until after the ready promise is created. (Otherwise it will run while playState is idle)
- Update tickAnimation so that it runs the auto align start time procedure
- Take range duration into account when
  - converting to and from CSSNuberish
  - calculating computed timing
</details>
<details>
<summary>Summary of changes to scroll-timeline-base:</summary>

- Remove renormalization callback, as it was only necessary when start and end delay was mapped to ranges
- Store additional values on new source and on resize: scrollWidth, scrollHeight, clientWidth, clientHeight
  - Values are stored once after resize
- Store ranges and update them on resize, as the auto align procedure will access them frequently
</details>

### CODE REVIEW

**Review by flackr — COMMENTED** [review]
Overall very cool to see the auto-aligned start time calculations implemented.

I just landed support for showing the diff on test results https://github.com/flackr/scroll-timeline/commit/5a2639adcd5691004177b10bd9970f4cea4b4fc0 . When you push another commit to this branch it should be enough to kick off the github action that will add the diff in the test results to the PR. I'm hoping this will help demonstrate that this is tested by the existing wpt tests and passes more as a result.

**Comment by johannesodland** [comment]
> Overall very cool to see the auto-aligned start time calculations implemented.
> 
> I just landed support for showing the diff on test results [5a2639a](https://github.com/flackr/scroll-timeline/commit/5a2639adcd5691004177b10bd9970f4cea4b4fc0) . When you push another commit to this branch it should be enough to kick off the github action that will add the diff in the test results to the PR. I'm hoping this will help demonstrate that this is tested by the existing wpt tests and passes more as a result.

Nice :)
I've been running tests and diffing by hand. It will be nice to have this automated 🙌

File: src/scroll-timeline-base.js [hunk 1]
```diff
@@ -184,47 +194,61 @@ function updateSource(timeline, source) {
...
           // "ResizeObserver loop completed with undelivered notifications"
           for (const timeline of details.timelines) {
-            renormalizeAnimationTimings(timeline);
+            updateInternal(timeline);
           }
         });
       });
       resizeObserver.observe(source);
+      for (const child of source.children) {
+        resizeObserver.observe(child);
```
[the remark this answers is not in the document]
Comment (johannesodland) [reply src/scroll-timeline-base.js:232]: It was necessary to pass some of the subtests, but we can definitely defer this.

File: src/scroll-timeline-base.js [hunk 2]
```diff
@@ -298,6 +313,9 @@ export class ScrollTimeline {
       // Internal members
       animations: [],
       scrollListener: null,
+
+      // Stored ranges
+      ranges: {}
```
[the remark this answers is not in the document]
Comment (johannesodland) [reply src/scroll-timeline-base.js:318]: That might be a cleaner solution, indeed. 

We just need to store some more measurements about the subject, such as top and left. `calculateRange` currently iterates through all offset parents up to the scroll containers offsetParent, calling offsetLeft and offsetTop to calculate the relative position of the subject. This will invalidate the layout.

I'll look into it and update the PR

File: test/expected.txt [hunk 3]
```diff
@@ -547,15 +547,15 @@ FAIL	/scroll-animations/css/view-timeline-shorthand.html	Shorthand contraction o
 FAIL	/scroll-animations/css/view-timeline-with-delay-and-range.tentative.html	ViewTimeline with animation delays and range
 FAIL	/scroll-animations/css/view-timeline-with-transform-on-subject.html	ViewTimeline use untransformed box for range calculations
 PASS	/scroll-animations/scroll-timelines/cancel-animation.html	Canceling an animation should cause its start time and hold time to be unresolved
-PASS	/scroll-animations/scroll-timelines/cancel-animation.html	A play-pending ready promise should be rejected when the animation is canceled
+FAIL	/scroll-animations/scroll-timelines/cancel-animation.html	A play-pending ready promise should be rejected when the animation is canceled
```
Comment (johannesodland) [inline test/expected.txt:550]: Adding the current results here for comparison. Will look into the failing tests.

File: src/scroll-timeline-base.js (hunk 1 above)
[the remark this answers is not in the document]
Comment (johannesodland) [reply src/scroll-timeline-base.js]: I have dropped the child observations. It doesn't cause any regressions, so that was definitely excessive 🙏

I have, however, added a mutation observer for class and style attributes to prevent regressions in /scroll-animations/scroll-timelines/scroll-animation-inactive-timeline.html.

I hope that's OK.

File: src/scroll-timeline-base.js (hunk 2 above)
[the remark this answers is not in the document]
Comment (johannesodland) [reply src/scroll-timeline-base.js:325]: It turns out that `calculateRange` is used in scroll-timeline-css.js to calculate keyframe offsets as well. It is used before any timelines have been instantiated, so we won't have access to any stored measurements here. 

https://github.com/flackr/scroll-timeline/blob/5a2639adcd5691004177b10bd9970f4cea4b4fc0/src/scroll-timeline-css.js#L96-L102

https://github.com/flackr/scroll-timeline/blob/5a2639adcd5691004177b10bd9970f4cea4b4fc0/src/scroll-timeline-css.js#L54-L58

We could refactor calculateRange to separate measurements and calculations, so that it can be used both places. 
The measurment and calculation of the subjects relative position should also be included in the stored measurements to prevent reflow:
https://github.com/flackr/scroll-timeline/blob/5a2639adcd5691004177b10bd9970f4cea4b4fc0/src/scroll-timeline-base.js#L522-L532

If we do this, the stored measurements will grow to include. 
```js
const measurements = {
  source: {
    scrollLeft,
    scrollTop,
    scrollWidth,
    scrollHeight,
    clientWidth,
    clientHeight,
    writingMode
  },
  subject: {
    top,
    left,
    offsetWidth,
    offsetHeight
  }
}
```

I'll start to look into it, hopefully next week.

[the remark this answers is not in the document]
Comment (flackr) [reply src/scroll-timeline-base.js:325]: This sounds good to me. I was expecting we would include all required measurements in the structure so that once we've stored it we never access any layout inducing properties again.

I have a longer term goal to do this for all timelines before setting the progress in any animations so that we don't dirty the layout from one timeline update before the next one does more calculations but we don't have to do that in this patch.

File: src/scroll-timeline-base.js [hunk 4]
```diff
@@ -181,50 +227,36 @@ function updateSource(timeline, source) {
...
-          }
-        });
-      });
+      // Use resize observer to detect changes to source size
+      const resizeObserver = new ResizeObserver(() => updateSourceMeasurements(source));
       resizeObserver.observe(source);
 
+      // Use mutation observer to detect updated style and class attributes on source element
+      const mutationObserver = new MutationObserver(() => updateSourceMeasurements(source));
+      mutationObserver.observe(source, {attributes: true, attributeFilter: ['style', 'class']});
```
Comment (flackr) [inline src/scroll-timeline-base.js:240]: Sorry to go back and forth on this - I think the ResizeObserver observing each child is better than observing changes to style / class attributes. It would be good to also have a mutationobserver observe added / removed children but this can be done separately.

File: test/expected.txt
Comment (flackr) [inline test/expected.txt:1]: This looks amazing! Can't wait to get this landed!

File: src/scroll-timeline-base.js (hunk 2 above)
[the remark this answers is not in the document]
Comment (johannesodland) [reply src/scroll-timeline-base.js:325]: 👍 

We are already storing source measurements and calculated ranges per source element (and not per timeline) before updating the timelines and ticking the animations. I don't think refactoring will change this, so we'll store both source measurements, subject measurements and relative positions, before we're updating the timelines and animations :)

Atm we're only observing changes to the source element. I'm not sure how we'll do this once we start observing the subject, but we'll look into this when we get there :)

File: src/scroll-timeline-base.js (hunk 4 above)
Comment (johannesodland) [reply src/scroll-timeline-base.js:240]: Don't be, I'm truly happy for all the input. 

We can definitely observe added/removed children. I think now is a good time to include that, but I can file a separate PR for that if you'd like that. 


About observing the style attribute:

There are a few tests that toggle whether the timelines are active through toggling the overflow style property.
- https://github.com/web-platform-tests/wpt/blob/55a2476161170898d9f19db51918353b2e2e9640/scroll-animations/scroll-timelines/scroll-animation-inactive-timeline.html#L61
- https://github.com/web-platform-tests/wpt/blob/55a2476161170898d9f19db51918353b2e2e9640/scroll-animations/scroll-timelines/scroll-animation-inactive-timeline.html#L89
 
There were regressions on a few of these tests, as we now have to wait for a startTime/holdTime before committing a pending play task.
 
I added the mutation observer to detect when the overflow property changed through inline styles. The approach is admittedly naive and triggers unnecessary updates, while only detecting a limited subset of the style changes we're interested in.

We can:
1. drop the observer and live with the few regressions, or
2. come up with a smarter way to detect the changes 🤔 Let me know if you have any good ideas 😀

Comment (flackr) [reply src/scroll-timeline-base.js:240]: > I can file a separate PR for that if you'd like that.

No need! I was worried about properly cleaning up listeners when children are removed / observing new children but the former is handled by the disconnect call, and the latter we can make part of the mutationobserver later.

> I added the mutation observer to detect when the overflow property changed through inline styles. The approach is admittedly naive and triggers unnecessary updates, while only detecting a limited subset of the style changes we're interested in.

Ah I understand, thanks for the explanation. I'm okay with keeping this in for now. For these things which are for handling edge cases correctly for tests but not required most of the time I'd like to eventually put them all behind some flag. E.g. if people are using this in production scenarios they would probably prefer not to pay any additional costs and instead carefully avoid those edge cases or manually trigger an update themselves.

Comment (johannesodland) [reply src/scroll-timeline-base.js]: Using a flag to toggle this would be nice. 

I'll keep observation of the "style" attribute then, but remove observation of the "class" attribute, as that's not necessary for passing the tests.

File: src/scroll-timeline-base.js (hunk 2 above)
[the remark this answers is not in the document]
Comment (johannesodland) [reply src/scroll-timeline-base.js]: @flackr I have refactored the code now to separate measurement and range calculations.

On resize and style changed callbacks we first measure the source element any timeline subjects, before updating timelines and animations. Ranges are now calculated on the fly based on the measurements.

I have also switched to using weak refs for the set of timelines that are connected to a source element, so that timelines are not prevented from being garbage-collected.

**Review by flackr — APPROVED** [review]
This looks great! Thank you! I'll land this before my PR so as to not create any conflicts needing resolving.
